dynamiqs/qarray.py

```diff
@@ -0,0 +1,535 @@
+        """Returns the check if the quantum state is a density matrix.
+
+        Returns:
+            True if the quantum state is a density matrix, False otherwise.
+        """
+
+    def isop(self) -> bool:
+        """Returns the check if the quantum state is an operator.
+
+        Returns:
+            True if the quantum state is an operator, False otherwise.
+        """
+        return self.isdm()
+
+    @abstractmethod
+    def isherm(self) -> bool:
+        """Returns the check if the quantum state is Hermitian.
+
+        Returns:
+            True if the quantum state is Hermitian, False otherwise.
+        """
+
+    @abstractmethod
+    def toket(self) -> QArray:
+        """Convert the quantum state to a ket.
+
+        Returns:
+            The ket representation of the quantum state.
+        """
+
+    @abstractmethod
+    def tobra(self) -> QArray:
+        """Convert the quantum state to a bra.
+
+        Returns:
+            The bra representation of the quantum state.
+        """
+
+    @abstractmethod
+    def todm(self) -> QArray:
+        """Convert the quantum state to a density matrix.
+
+        Returns:
+            The density matrix representation of the quantum state.
+        """
+
+    def toop(self) -> QArray:
+        """Convert the quantum state to an operator.
+
+        Returns:
+            The operator representation of the quantum state.
+        """
+        return self.todm()
+
+    def proj(self) -> QArray:
+        """Projector of the quantum state.
+
+        Returns:
+            The projector of the quantum state.
+        """
+        return self.todm()
+
+    # conversion methods
+    @abstractmethod
+    def to_numpy(self) -> np.ndarray:
+        """Convert the quantum state to a NumPy array.
+
+        Returns:
+            The NumPy array representation of the quantum state.
+        """
+
+    @abstractmethod
+    def to_qutip(self) -> qt.QObj:
+        """Convert the quantum state to a QuTiP object.
+
+        Returns:
+            A QuTiP object representation of the quantum state.
+        """
+
+    @abstractmethod
+    def to_jax(self) -> Array:
+        """Convert the quantum state to a JAX array.
+
+        Returns:
+            The JAX array representation of the quantum state.
+        """
+
+    def __repr__(self) -> str:
+        return (
+            f'{type(self).__name__}(shape={self.shape}, '
+            f'dims={self.dims}, dtype={self.dtype})'
+        )
+
+    def __neg__(self) -> QArray:
+        """Negate the quantum state."""
+        return -1 * self
+
+    @abstractmethod
+    def __mul__(
+        self, other: ScalarLike | ArrayLike
+    ) -> QArray:  # warning if used with array
+        """Element-wise multiplication with a scalar or an array."""
+
+    def __rmul__(
+        self, other: ScalarLike | ArrayLike
+    ) -> QArray:  # warning if used with array
+        """Element-wise multiplication with a scalar or an array on the right."""
```

Comment:
Let's only authorize `ScalarLike` for these two (as we did in timearray) to simplify subclasses implementations.